kernel/virtio/virtio_net.c

```diff
@@ -203,38 +203,40 @@ static void recv_setup(void)
 int virtio_net_xmit_packet(void *data, int len)
 {
     struct virtq_desc *desc;
-    struct virtq_avail *avail;
     int dbg = 0;
 
-    if (((xmit_next_avail + 1) % xmitq.num) ==
-        (xmit_last_used % xmitq.num)) {
+    if (((xmit_next_avail + 2) % xmitq.num) ==
+        ((xmit_last_used * 2) % xmitq.num)) {
```

Comment:
That's because the `tx_buf`'s indexes used for data match `xmit_next_avail`. So, the tx_bufs look like this: empty, data, empty, data, empty, data. It should look better with the change you suggest below: getting an xmit_buf for the header as well.

Btw, my plan is to make the xmit_bufs dynamic as well (next PR). Just malloc that memory and free it on the interrupt side.